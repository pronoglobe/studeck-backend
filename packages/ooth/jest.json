{
  "collectCoverage": true,
  "transform": {
    "^.+\\.tsx?$": "ts-jest"
  },
  "testEnvironment": "node",
  "testMatch": ["**/test/**/*.test.ts"],
  "testPathIgnorePatterns": ["/node_modules/"],
  "moduleFileExtensions": ["ts", "tsx", "js", "json"]
}
