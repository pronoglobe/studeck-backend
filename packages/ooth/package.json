{
  "name": "ooth",
  "version": "2.12.0",
  "description": "",
  "main": "lib/index.js",
  "scripts": {
    "clean": "del-cli ./lib ./coverage",
    "build": "npm run clean && tsc -p ./config/tsconfig.build.json",
    "lint": "tslint -c ./tslint.json -p ./config/tsconfig.build.json",
    "test:integration": "jest -c ./jest.json",
    "test": "npm run lint && npm run clean && npm run test:integration",
    "test:watch": "npm run clean && jest -c ./jest.json --watch"
  },
  "author": "Nick Redmark <nicolamr@gmail.com> (http://nmr.io)",
  "license": "MIT",
  "engines": {
    "node": ">=8"
  },
  "repository": "nmaro/ooth",
  "dependencies": {
    "body-parser": "^1.16.0",
    "cookie-parser": "^1.4.3",
    "express": "^4.14.1",
    "express-session": "^1.15.6",
    "locale": "^0.1.0",
    "ooth-i18n": "^2.12.0",
    "passport": "^0.3.2",
    "passport-custom": "^1.0.5"
  },
  "devDependencies": {
    "@smartive/tslint-config": "^4.0.0",
    "@types/body-parser": "^1.17.0",
    "@types/cookie-parser": "^1.4.1",
    "@types/express-session": "^1.15.10",
    "@types/jest": "^23.3.5",
    "@types/lodash": "^4.14.117",
    "@types/mongodb": "^3.1.4",
    "@types/passport": "^0.4.6",
    "@types/passport-jwt": "^3.0.1",
    "@types/request-promise": "^4.1.42",
    "del-cli": "^1.1.0",
    "jest": "^23.5.0",
    "lodash": "^4.17.10",
    "mongodb": "^3.1.3",
    "mongodb-memory-server": "^2.0.1",
    "ms": "^2.1.1",
    "ooth-mongo": "^2.12.0",
    "passport-strategy": "^1.0.0",
    "request-promise": "^4.2.1",
    "ts-jest": "^23.1.4",
    "tslint": "^5.11.0",
    "tslint-config-prettier": "^1.15.0",
    "tsutils": "^3.0.0",
    "typescript": "^3.0.1"
  }
}
