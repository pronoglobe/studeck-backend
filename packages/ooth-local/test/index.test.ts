import * as express from 'express';
import * as session from 'express-session';
import { cloneDeep } from 'lodash';
import { MongoClient, Db } from 'mongodb';
import MongodbMemoryServer from 'mongodb-memory-server';
import { Ooth } from 'ooth';
import emailer from 'ooth-local-emailer';
import { OothMongo } from 'ooth-mongo';
import * as request from 'request-promise';

import oothLocal from '../src';
import { Server } from 'http';

const { Strategy } = require('passport-custom');

let mongoServer: MongodbMemoryServer;
let con: MongoClient;
let app: express.Express;
let server: Server;
let ooth: Ooth;
let db: Db;
let cookies = '';
let oothMongo: OothMongo;

const startServer = () => {
  return new Promise((resolve) => {
    server = app.listen(8080, resolve);
  });
};

const obfuscate = (obj: any, ...paths: string[]) => {
  const res = cloneDeep(obj);
  for (const path of paths) {
    const keys = path.split('.');
    let current = res;
    for (const key of keys.slice(0, -1)) {
      current = current[key];
    }
    current[keys[keys.length - 1]] = '<obfuscated>';
  }

  return res;
};

const obfuscatePatterns = (obj: any, ...patterns: RegExp[]): any => {
  if (typeof obj === 'object') {
    const res: any = {};
    for (const key of Object.keys(obj)) {
      res[key] = obfuscatePatterns(obj[key], ...patterns);
    }
    return res;
  }

  if (Array.isArray(obj)) {
    return obj.map((i) => obfuscatePatterns(i, ...patterns));
  }

  if (typeof obj === 'string') {
    return patterns.reduce((s, pattern) => s.replace(pattern, '<obfuscated>'), obj);
  }

  return obj;
};

describe('ooth-local', () => {
  let onForgotPasswordListener: any;
  let onRequestVerifyListener: any;
  let onRegisterListener: any;
  let userId: string;
  let resetToken: string;
  let verificationToken: string;
  let sendMail: any = () => null;

  beforeAll(async () => {
    mongoServer = new MongodbMemoryServer();
    const connectionString = await mongoServer.getConnectionString();
    const dbName = await mongoServer.getDbName();
    con = await MongoClient.connect(connectionString);
    db = con.db(dbName);
  });

  afterAll(async () => {
    await con.close();
    await mongoServer.stop();
  });

  beforeEach(async () => {
    app = express();
    app.use(
      session({
        name: 'api-session-id',
        secret: 'x',
        resave: false,
        saveUninitialized: true,
      }),
    );
    oothMongo = new OothMongo(db);
    ooth = new Ooth({
      app,
      backend: oothMongo,
      path: '',
    });
    oothLocal({
      ooth,
      translations: {
        en: require('../i18n/en.json'),
        fr: {
          register: {
            invalid_email: 'french invalid email',
          },
        },
      },
    });
    sendMail = jest.fn();
    emailer({
      ooth,
      from: 'noreply@example.com',
      siteName: 'Example Site',
      sendMail: (...args) => sendMail(...args),
    });
    ooth.on('local', 'register', async (...args) => {
      if (onRegisterListener) {
        onRegisterListener(...args);
      }
    });
    ooth.on('local', 'forgot-password', async (data) => {
      if (onForgotPasswordListener) {
        //Store user id and reset token for follow on test
        userId = data._id;
        resetToken = data.passwordResetToken;
        onForgotPasswordListener(data);
      }
    });
    ooth.on('local', 'generate-verification-token', async (data) => {
      if (onRequestVerifyListener) {
        //Store user id and  verification token for follow on test
        userId = data._id;
        verificationToken = data.verificationToken;
        onRequestVerifyListener(data);
      }
    });
    ooth.registerAfterware(async (res, userId) => {
      if (userId) {
        res.user = ooth.getProfile(await ooth.getUserById(userId));
      }

      return res;
    });
    await startServer();
  });

  afterEach(async () => {
    if (server) {
      await server.close();
    }
    if (db) {
      await db.dropDatabase();
    }
    cookies = '';
  });

  test('fails without email', async () => {
    try {
      await request({
        method: 'POST',
        uri: 'http://localhost:8080/local/register',
        json: true,
      });
    } catch (e) {
      expect(e.response.body).toMatchSnapshot();
      return;
    }
    throw new Error("Didn't fail");
  });

  test('translates', async () => {
    try {
      await request({
        method: 'POST',
        uri: 'http://localhost:8080/local/register',
        headers: {
          'Accept-Language': 'fr',
        },
        json: true,
      });
    } catch (e) {
      expect(e.response.body).toMatchSnapshot();
      return;
    }
    throw new Error("Didn't fail");
  });

  test('fails without password', async () => {
    try {
      await request({
        method: 'POST',
        uri: 'http://localhost:8080/local/register',
        body: {
          email: 'test@example.com',
        },
        json: true,
      });
    } catch (e) {
      expect(e.response.body).toMatchSnapshot();
      return;
    }
    throw new Error("Didn't fail");
  });

  test('fails without valid password', async () => {
    try {
      await request({
        method: 'POST',
        uri: 'http://localhost:8080/local/register',
        body: {
          email: 'test@example.com',
          password: 'xxxxxx',
        },
        json: true,
      });
    } catch (e) {
      expect(e.response.body).toMatchSnapshot();
      return;
    }
    throw new Error("Didn't fail");
  });

  test('can register', async () => {
    onRegisterListener = jest.fn();
    const res = await request({
      method: 'POST',
      uri: 'http://localhost:8080/local/register',
      body: {
        email: 'test@example.com',
        password: 'Asdflba09',
      },
      json: true,
    });
    expect(res).toMatchSnapshot();
    expect(obfuscate(onRegisterListener.mock.calls[0][0], '_id', 'verificationToken')).toMatchSnapshot();
    expect(obfuscatePatterns(sendMail.mock.calls, /token=[\w\d]+/g, /userId=[\w\d]+/g)).toMatchSnapshot();
  });

  describe('after registration', () => {
    beforeEach(async () => {
      await request({
        method: 'POST',
        uri: 'http://localhost:8080/local/register',
        body: {
          email: 'test@example.com',
          password: 'Asdflba09',
        },
        json: true,
      });
      sendMail = jest.fn();
    });

    test('login fails with wrong password', async () => {
      try {
        await request({
          method: 'POST',
          uri: 'http://localhost:8080/local/login',
          body: {
            username: 'test@example.com',
            password: 'Asdflba099',
          },
          json: true,
        });
      } catch (e) {
        expect(e.response.body).toMatchSnapshot();
        return;
      }
      throw new Error("Didn't fail");
    });

    test('can login', async () => {
      const res = await request({
        method: 'POST',
        uri: 'http://localhost:8080/local/login',
        body: {
          username: 'test@example.com',
          password: 'Asdflba09',
        },
        json: true,
      });
      delete res.user._id;
      expect(res).toMatchSnapshot();
    });

    test('can forget password', async () => {
      onForgotPasswordListener = jest.fn();
      await request({
        method: 'POST',
        uri: 'http://localhost:8080/local/forgot-password',
        body: {
          username: 'test@example.com',
        },
        json: true,
      });
      expect(obfuscate(onForgotPasswordListener.mock.calls[0][0], '_id', 'passwordResetToken')).toMatchSnapshot();
      expect(obfuscatePatterns(sendMail.mock.calls, /token=[\w\d]+/g, /userId=[\w\d]+/g)).toMatchSnapshot();
    });

    test('can reset password', async () => {
      onForgotPasswordListener = jest.fn();

      //Store the password reset token and userId so we can use it later
      await request({
        method: 'POST',
        uri: 'http://localhost:8080/local/forgot-password',
        body: {
          username: 'test@example.com',
        },
        json: true,
      });
      sendMail = jest.fn();
      const res = await request({
        method: 'POST',
        uri: 'http://localhost:8080/local/reset-password',
        body: {
          userId,
          token: resetToken,
          newPassword: 'Asdflba10',
        },
        json: true,
      });
      expect(res).toMatchSnapshot();
      expect(obfuscatePatterns(sendMail.mock.calls, /token=[\w\d]+/g, /userId=[\w\d]+/g)).toMatchSnapshot();
    });

    test("can't reset password with invalid token", async () => {
      onForgotPasswordListener = jest.fn();

      //Store the password reset token and userId so we can use it later
      await request({
        method: 'POST',
        uri: 'http://localhost:8080/local/forgot-password',
        body: {
          username: 'test@example.com',
        },
        json: true,
      });
      try {
        await request({
          method: 'POST',
          uri: 'http://localhost:8080/local/reset-password',
          body: {
            userId,
            token: 'wrongtoken',
            newPassword: 'Asdflba10',
          },
          json: true,
        });
      } catch (e) {
        expect(e.response.body).toMatchSnapshot();
        return;
      }
      throw new Error("Didn't fail");
    });

    describe('after login', () => {
      beforeEach(async () => {
        sendMail = jest.fn();
        const res = await request({
          method: 'POST',
          uri: 'http://localhost:8080/local/login',
          body: {
            username: 'test@example.com',
            password: 'Asdflba09',
          },
          json: true,
        });
        ooth.registerSecondaryAuth(
          'foo',
          'bar',
          () => true,
          new Strategy((_req: any, done: any) => done(null, res.user._id)),
        );
      });

      test('can generate verification token', async () => {
        onRequestVerifyListener = jest.fn();
        await request({
          method: 'POST',
          uri: 'http://localhost:8080/local/generate-verification-token',
          json: true,
          headers: {
            Cookie: cookies,
          },
        });
        expect(obfuscate(onRequestVerifyListener.mock.calls[0][0], '_id', 'verificationToken')).toMatchSnapshot();
        expect(obfuscatePatterns(sendMail.mock.calls, /token=[\w\d]+/g, /userId=[\w\d]+/g)).toMatchSnapshot();
      });

      test('can verify user with token', async () => {
        onRequestVerifyListener = jest.fn();
        //Store verification token and userId for later use
        await request({
          method: 'POST',
          uri: 'http://localhost:8080/local/generate-verification-token',
          json: true,
          headers: {
            Cookie: cookies,
          },
        });
        sendMail = jest.fn();
        const res = await request({
          method: 'POST',
          uri: 'http://localhost:8080/local/verify',
          json: true,
          body: {
            userId,
            token: verificationToken,
          },
          headers: {
            Cookie: cookies,
          },
        });
        expect(obfuscate(res, 'user._id')).toMatchSnapshot();
        expect(obfuscatePatterns(sendMail.mock.calls, /token=[\w\d]+/g, /userId=[\w\d]+/g)).toMatchSnapshot();
      });

      test("can't verify user with invalid token", async () => {
        onRequestVerifyListener = jest.fn();
        //Store verification token and userId for later use
        await request({
          method: 'POST',
          uri: 'http://localhost:8080/local/generate-verification-token',
          json: true,
          headers: {
            Cookie: cookies,
          },
        });
        try {
          await request({
            method: 'POST',
            uri: 'http://localhost:8080/local/verify',
            json: true,
            body: {
              userId,
              token: 'wrongtoken',
            },
            headers: {
              Cookie: cookies,
            },
          });
        } catch (e) {
          expect(e.response.body).toMatchSnapshot();
          return;
        }
        throw new Error("Didn't fail");
      });

      test('can set username', async () => {
        const res = await request({
          method: 'POST',
          uri: 'http://localhost:8080/local/set-username',
          body: {
            username: 'heythere12_',
          },
          json: true,
          headers: {
            Cookie: cookies,
          },
        });
        delete res.user._id;
        expect(res).toMatchSnapshot();
      });

      test('can set email', async () => {
        const res = await request({
          method: 'POST',
          uri: 'http://localhost:8080/local/set-email',
          body: {
            email: 'test@example.com',
          },
          json: true,
          headers: {
            Cookie: cookies,
          },
        });
        delete res.user._id;
        expect(res).toMatchSnapshot();
        expect(obfuscatePatterns(sendMail.mock.calls, /token=[\w\d]+/g, /userId=[\w\d]+/g)).toMatchSnapshot();
      });

      test('can change password', async () => {
        const res = await request({
          method: 'POST',
          uri: 'http://localhost:8080/local/change-password',
          body: {
            password: 'Asdflba09',
            newPassword: 'XXAsdflba09',
          },
          json: true,
          headers: {
            Cookie: cookies,
          },
        });
        delete res.user._id;
        expect(res).toMatchSnapshot();
        expect(obfuscatePatterns(sendMail.mock.calls, /token=[\w\d]+/g, /userId=[\w\d]+/g)).toMatchSnapshot();
      });

      test("can't set invalid username", async () => {
        try {
          await request({
            method: 'POST',
            uri: 'http://localhost:8080/local/set-username',
            body: {
              username: 'bl',
            },
            json: true,
            headers: {
              Cookie: cookies,
            },
          });
        } catch (e) {
          expect(e.response.body).toMatchSnapshot();
          return;
        }
        throw new Error("Didn't fail");
      });
    });
  });
});
