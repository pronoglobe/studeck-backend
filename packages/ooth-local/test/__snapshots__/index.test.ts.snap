// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ooth-local after registration after login can change password 1`] = `
Object {
  "message": "Password has been changed.",
  "user": Object {
    "local": Object {
      "email": "test@example.com",
    },
  },
}
`;

exports[`ooth-local after registration after login can change password 2`] = `
Object {
  "0": Object {
    "0": Object {
      "body": "Your password for Example Site has been changed.",
      "from": "noreply@example.com",
      "html": "Your password for Example Site has been changed.",
      "subject": "Example Site password changed",
      "to": "test@example.com",
    },
  },
}
`;

exports[`ooth-local after registration after login can generate verification token 1`] = `
Object {
  "_id": "<obfuscated>",
  "email": "test@example.com",
  "verificationToken": "<obfuscated>",
}
`;

exports[`ooth-local after registration after login can generate verification token 2`] = `
Object {
  "0": Object {
    "0": Object {
      "body": "Please verify your Example Site email by opening the following url: /verify-email?<obfuscated>&<obfuscated>.",
      "from": "noreply@example.com",
      "html": "Please verify your Example Site email by opening the following url: /verify-email?<obfuscated>&<obfuscated>.",
      "subject": "Verify your Example Site address",
      "to": "test@example.com",
    },
  },
}
`;

exports[`ooth-local after registration after login can set email 1`] = `
Object {
  "message": "Email updated.",
  "user": Object {
    "local": Object {
      "email": "test@example.com",
    },
  },
}
`;

exports[`ooth-local after registration after login can set email 2`] = `
Object {
  "0": Object {
    "0": Object {
      "body": "Please verify your Example Site email by opening the following url: /verify-email?<obfuscated>&<obfuscated>.",
      "from": "noreply@example.com",
      "html": "Please verify your Example Site email by opening the following url: /verify-email?<obfuscated>&<obfuscated>.",
      "subject": "Verify your Example Site address",
      "to": "test@example.com",
    },
  },
}
`;

exports[`ooth-local after registration after login can set username 1`] = `
Object {
  "message": "Username updated.",
  "user": Object {
    "local": Object {
      "email": "test@example.com",
      "username": "heythere12_",
    },
  },
}
`;

exports[`ooth-local after registration after login can verify user with token 1`] = `
Object {
  "message": "Email verified",
  "user": Object {
    "_id": "<obfuscated>",
    "local": Object {
      "email": "test@example.com",
      "verified": true,
    },
  },
}
`;

exports[`ooth-local after registration after login can verify user with token 2`] = `
Object {
  "0": Object {
    "0": Object {
      "body": "Your Example Site address has been verified.",
      "from": "noreply@example.com",
      "html": "Your Example Site address has been verified.",
      "subject": "Example Site address verified",
      "to": "test@example.com",
    },
  },
}
`;

exports[`ooth-local after registration after login can't set invalid username 1`] = `
Object {
  "message": "Username must be all lowercase, contain only letters, numbers and _ (starting with a letter), and be between 4 and 20 characters long.",
  "status": "error",
}
`;

exports[`ooth-local after registration after login can't verify user with invalid token 1`] = `
Object {
  "message": "Verification token invalid, expired or already used.",
  "status": "error",
}
`;

exports[`ooth-local after registration can forget password 1`] = `
Object {
  "_id": "<obfuscated>",
  "email": "test@example.com",
  "passwordResetToken": "<obfuscated>",
}
`;

exports[`ooth-local after registration can forget password 2`] = `
Object {
  "0": Object {
    "0": Object {
      "body": "Reset your password for Example Site on the following page: /reset-password?<obfuscated>&<obfuscated>",
      "from": "noreply@example.com",
      "html": "Reset your password for Example Site on the following page: /reset-password?<obfuscated>&<obfuscated>",
      "subject": "Reset Example Site password",
      "to": "test@example.com",
    },
  },
}
`;

exports[`ooth-local after registration can login 1`] = `
Object {
  "user": Object {
    "local": Object {
      "email": "test@example.com",
    },
  },
}
`;

exports[`ooth-local after registration can reset password 1`] = `
Object {
  "message": "Password has been reset.",
}
`;

exports[`ooth-local after registration can reset password 2`] = `
Object {
  "0": Object {
    "0": Object {
      "body": "Your password for Example Site has been reset.",
      "from": "noreply@example.com",
      "html": "Your password for Example Site has been reset.",
      "subject": "Example Site password reset",
      "to": "test@example.com",
    },
  },
}
`;

exports[`ooth-local after registration can't reset password with invalid token 1`] = `
Object {
  "message": "Password reset token invalid, expired or already used.",
  "status": "error",
}
`;

exports[`ooth-local after registration login fails with wrong password 1`] = `
Object {
  "message": "Incorrect password.",
  "status": "error",
}
`;

exports[`ooth-local can register 1`] = `
Object {
  "message": "User registered successfully.",
}
`;

exports[`ooth-local can register 2`] = `
Object {
  "_id": "<obfuscated>",
  "email": "test@example.com",
  "verificationToken": "<obfuscated>",
}
`;

exports[`ooth-local can register 3`] = `
Object {
  "0": Object {
    "0": Object {
      "body": "Thank you for joining Example Site!",
      "from": "noreply@example.com",
      "html": "Thank you for joining Example Site!",
      "subject": "Welcome to Example Site",
      "to": "test@example.com",
    },
  },
  "1": Object {
    "0": Object {
      "body": "Please verify your Example Site email by opening the following url: /verify-email?<obfuscated>&<obfuscated>.",
      "from": "noreply@example.com",
      "html": "Please verify your Example Site email by opening the following url: /verify-email?<obfuscated>&<obfuscated>.",
      "subject": "Verify your Example Site address",
      "to": "test@example.com",
    },
  },
}
`;

exports[`ooth-local fails without email 1`] = `
Object {
  "message": "Invalid email",
  "status": "error",
}
`;

exports[`ooth-local fails without password 1`] = `
Object {
  "message": "Invalid password",
  "status": "error",
}
`;

exports[`ooth-local fails without valid password 1`] = `
Object {
  "message": "Password must contain digits, lowercase and uppercase letters and be at least 6 characters long.",
  "status": "error",
}
`;

exports[`ooth-local translates 1`] = `
Object {
  "message": "Invalid email",
  "status": "error",
}
`;
